File: source/go/dining_philosophers.go
package main

import (
	"fmt"
	"sync"
)

func philosopher(id int, forks []chan int, numPhil int, resultsCh chan<- int, wg *sync.WaitGroup) {
	defer wg.Done()

	left := id
	right := (id + 1) % numPhil

	first, second := left, right
	if left > right {
		first, second = right, left
	}

	// Pick up forks in order
	<-forks[first]
	<-forks[second]

	// Eat
	fmt.Printf("philosopher %d is eating\n", id)

	// Put down forks
	forks[first] <- 1
	forks[second] <- 1

	resultsCh <- id
}

func main() {
	numPhil := 5
	forks := make([]chan int, numPhil)
	for i := 0; i < numPhil; i++ {
		forks[i] = make(chan int, 1)
		forks[i] <- 1 // token = fork is available
	}
	resultsCh := make(chan int, numPhil)

	var wg sync.WaitGroup
	for i := 0; i < numPhil; i++ {
		wg.Add(1)
		go philosopher(i, forks, numPhil, resultsCh, &wg)
	}

	wg.Wait()

	var ids []int
	for i := 0; i < numPhil; i++ {
		ids = append(ids, <-resultsCh)
	}
	close(resultsCh)

	// Drain fork channels
	for i := 0; i < numPhil; i++ {
		<-forks[i]
		close(forks[i])
	}

	_ = ids

	fmt.Println("all philosophers ate")
}

File: source/c/dining_philosophers.c
// gcc -o dining_philosophers dining_philosophers.c -lpthread
#include <stdio.h>
#include <pthread.h>

#define NUM_PHIL 5

pthread_mutex_t forks[NUM_PHIL];

void *philosopher(void *arg) {
    int id = *(int*)arg;
    int left = id, right = (id + 1) % NUM_PHIL;
    int first = left < right ? left : right;
    int second = left < right ? right : left;

    pthread_mutex_lock(&forks[first]);
    pthread_mutex_lock(&forks[second]);

    printf("philosopher %d is eating\n", id);

    pthread_mutex_unlock(&forks[second]);
    pthread_mutex_unlock(&forks[first]);
    return NULL;
}

int main(void) {
    for (int i = 0; i < NUM_PHIL; i++) pthread_mutex_init(&forks[i], NULL);

    pthread_t tasks[NUM_PHIL];
    int ids[NUM_PHIL];
    for (int i = 0; i < NUM_PHIL; i++) {
        ids[i] = i;
        pthread_create(&tasks[i], NULL, philosopher, &ids[i]);
    }
    for (int i = 0; i < NUM_PHIL; i++) pthread_join(tasks[i], NULL);

    printf("all philosophers ate\n");
    return 0;
}
